### ISSUE TITLE
configuring root routes in the in the component

### ISSUE THREAD

**Issue by winnemucca** [angular/angular.js#13780]
Please direct me to another issue if this has been asked and resolved. 

I am working to get more accustomed to using .component and 'ngComponentRouter'.

I have set a component on that would serve as the root.  

```js
   (function() {
		'use strict;'
		angular
			.module('componentTutorial.home')
			.value('$routerRootComponent', 'componentTutorial.home')
			.component('home',HomeComponentOptions);

			var HomeComponentOptions = {
				// $routeConfig: [
				// 	{path: '/home...', name: 'Home', useAsDefault: true}
				// ],
				bindings: {  
					title: '@'
				},
				templateUrl: 'app/home/home.html',
				controller: HomeController	    
			}

			function HomeController ($router) {
				$routeConfig: [
					{path: '/home...', name: 'Home', useAsDefault: true}
				];
			}
	})();
```

However, I am seeing that that the new component router is asking for a root router with an error.

    Error: Component "Root" has no route config.

$routerRootComponent is located in the source file for the router.

  var router = new RootRouter(registry, location, $routerRootComponent);

However, there seems to be an issue reading it.

**Comment by petebacondarwin** [comment]
I think that (at the moment) since there is no concept of a "root component" in Angular 1 (there is in Angular 2) you have to configure the root component directly in a component router service.

**Comment by Sriram143** [comment]
use router.config() function

function HomeController ($router) {
                $router.config([
                    {path: '/home...', name: 'Home', useAsDefault: true}
                ]);
            }

**Comment by petebacondarwin** [comment]
Yes, that sort of thing. Thanks @Sriram143

**Comment by winnemucca** [comment]
@petebacondarwin @Sriram143.   

I recognize $router.config from some previous tutorials posted online.  However, that does not change the error message that is posted.   I declared the *$routerRootComponent* in a value service. and changed that to the desired component name.   However that did not change the errors but added additional errors  

    `Cannot read property '$canActivate' of undefined `

Here is the code.  I believe that I am following configuring it in an appropriate way.  

    (function() {
		'use strict;'
		angular
			.module('componentTutorial')
			.value('$routerRootComponent', 'HomeComponentOptions')
			.component('home',HomeComponentOptions);

			function defaultCanActivateRouterHook(componentName) {
	            return function() {
	                console.log("${componentName}: $canActivate called with", componentName);
	                return true;
	            }
        	}
        
	        function defaultRouterHooks(ctrlName) {
	            return {
	                $routerOnActivate: function(next, prev) {
	                    console.log( next, prev);
	                },
	                $routerOnDeactivate: function (next, prev)  {
	                    console.log( next, prev);
	                },              
	                $routerOnReuse: function (next, prev) {
	                    console.log(next, prev);
	                },      
	                $routerCanReuse: function (next, prev)  {
	                    console.log( next, prev);
	                    return true;
	                },             
	                $routerCanDeactivate: function (next, prev) {
	                    console.log( next, prev);
	                    return true
	                }
	            }
	        }
        
			var HomeComponentOptions = {
				
				bindings: {  
					title: '@'
				},
				// $canActivate: defaultCanActivateRouterHook('DempAppComponent'),

				templateUrl: 'app/home/home.html',
				controller: HomeController	    
			}


			function HomeController ($router) {
				$router.config([
					{path: '/home...', name: 'Home', useAsDefault: true}
				]);
			}
		
	})();

**Comment by gkalpak** [comment]
I don't think `component()` is compatible with the new router yet, because components are restricted to elements, while the router expects attribute directives/components. This will be fixed soon.

**Comment by winnemucca** [comment]
@petebacondarwin 
I deleted my last response since considering assuming it was too much to read.   I am still trying to understand the root component and figuring everything off of that.   My question is if a root rout can be configured?

I have set a service value based on what I can search on github and have read in the router files.

    	.value('$routerRootComponent', 'demoApp')

I set 'demoApp' as my root component and am using in the manner as advised.   Following @Sriram143 advice I changed routeConfig to $router.config.   The initial use **useAsDefault** works, but I do not see if components can be accessed inside of a 

    <ng-outlet="main"> </ng-outlet>

I set a $canActivate to my issue and removed root errors but accessibilities to routers has not changed.  Is this something I must wait on?  Is this a issue being resolved?

**Comment by petebacondarwin** [comment]
@winnemucca I haven't played with the router for a while so I can't answer this immediately. If @btford doesn't comment in the next few days I will get on top of it early next week.

**Comment by geroyche** [comment]
it's actually quite confusing.
the comments in angular_1_router.js state

 * The directive uses a simple syntax: `ng-link="componentName({ param: paramValue })"`

when following that syntax, i get this error:
    TypeError: Cannot read property 'reduce' of undefined
    at splitAndFlattenLinkParams (angular_1_router.js:2332)
    at RouteRegistry.generate (angular_1_router.js:2129)
    at RootRouter.Router.generate (angular_1_router.js:2753)
    at getLink (angular_1_router.js:227)

after some searching i found that actually the syntax is now probably rather
ng-link=[/componentName', {guuid: 'none'}]

using the later syntax i get the error message discussed here, and I fail to set 
`.value('$routerRootComponent', 'app');`
as well

**Comment by brandonroberts** [comment]
@geroyche here is a plunker with an example of setting a root component and a couple of route links: http://plnkr.co/edit/bzmgUygZ5nPLDaU7ucHB?p=preview

**Comment by geroyche** [comment]
@brandonroberts thanks!
here's a rewrite of it though, trying to use components rather than directives, that fails with the router in the manner described above
http://plnkr.co/edit/ZnvsKeBgTzmufAZofdei?p=preview

i generally understood that directives will/should be abandoned in favor of components ?

**Comment by geroyche** [comment]
ok. so i see how it works well when using .directive() here: https://github.com/brandonroberts/angularjs-component-router/tree/master/app/components

i guess now the question is, whether all the documentation on the abandoned angular-new-router
( https://angular.github.io/router/getting-started ), which i wasted a couple of days with before finding angular_1_router.js. is outdated, and .components() is not really the way to go anymore.

**Comment by gkalpak** [comment]
The reason why it fails has already been explained in https://github.com/angular/angular.js/issues/13780#issuecomment-172828417.
It's a temp issue though :smiley:

**Comment by geroyche** [comment]
hmmm. that leaves me between a rock and a hard place in my current project.
either stick to angular-new-router for the time being (which seems to lack proper support for child routes), or refactor the exisiting components to directives for the time being.
but okay, noted.

**Comment by brandonroberts** [comment]
I don't think you should stick with the angular-new-router either way. You could use directives for your routes for the time being and use components for everything else. Or you could stay on 1.5.0.rc0 which still allows for element and attribute components.

**Comment by petebacondarwin** [comment]
We are going to fix all this up, this week.

**Comment by winnemucca** [comment]
@petebacondarwin @brandonroberts  @btford  Where can we go to keep up with updated ngComponentRouter files?

**Comment by petebacondarwin** [comment]
@winnemucca - that is a good question. You can start following these files: https://github.com/angular/angular/tree/master/modules/angular1_router

**Comment by petebacondarwin** [comment]
Here is an example application of using the component router with ng1 - http://plnkr.co/edit/N3YP3dKMuljpZ6mWsVBT?p=preview
Docs and guide to follow

**Comment by winnemucca** [comment]
Super pumped!   Already shared this with my team.  Huge thanks!!!

**Comment by petebacondarwin** [comment]
Oh and there is a repository with the demo app too: https://github.com/petebacondarwin/ng1-component-router-demo

**Comment by petebacondarwin** [comment]
And here is a draft of the guide (WIP): https://docs.google.com/document/d/1vl8_mA7rBSTVnPUMDiPpM37OPeM81SypKVvbBSuISgw/

**Comment by brandonroberts** [comment]
@petebacondarwin are the lifecycle hooks the only way you get access to route params and route data?

**Comment by petebacondarwin** [comment]
With the current published router the only way to get hold of the router and params is to "require" the `ngOutlet` directive in your component. But I hacked it to be able to get it from the activate hook.

**Comment by brandonroberts** [comment]
Ok

**Comment by winnemucca** [comment]
Hey @petebacondarwin I am closing this issue thank you!  I am wondering though if/how to add multiple views with  

``` {path: '/crisis-center/...', name: 'CrisisCenter', components:{main: 'crisisCenter', secondary: 'randomInfo' }, useAsDefault: true }, ```


``` <div class="col-xs-6 col-sm-9 col-md-9">
   		<ng-outlet ="main"></ng-outlet>
    </div>

    <div class="col xs-6 col-sm -9">
    	<h1> Another View</h1>
    		<ng-outlet="secondary"> </ng-outlet>
    </div>```

**Comment by petebacondarwin** [comment]
@winnemucca this is not currently available in the Component Router.

**Comment by winnemucca** [comment]
Cool thanks!

**Comment by Rab815** [comment]
@petebacondarwin I've been trying to get an example much like your 
http://plnkr.co/edit/N3YP3dKMuljpZ6mWsVBT?p=preview
working except instead of using templates and everything in one file the way you have it but actually having the components using.html files in a directory structure such as heroes/ , heroeslist, and heroesdetails components are in a nested directory  such as app/components/heroes  where the heroes.js, herosList.js and herosDetails.js are separate files each with an associated html file to keep templates out of js code  but the pages come up blank when just adjusting the plunker to use html templateUrls instead of just inline template: is the ngComponentRouter supposed to automatically find the location of the html templates?  Sorry for all the questions just trying to wrap my head around this stuff

**Comment by Rab815** [comment]
also how are you injecting the content of <ng-outlet> as an element ie... <heroes> vs a <div heroes> div with a heroes attribute

**Comment by petebacondarwin** [comment]
It should be in the latest router builds

**Comment by petebacondarwin** [comment]
What about when you get it from npm?

**Comment by Rab815** [comment]
I got the example working. Brandon. Roberts told me the $route to $rootRouter change hasn't made it to master yet.   I've broken it up to have separate html files expecting it to automatically find them based on component name however that didn't work. Apparently $componentLoaderProvider override is no longer available

**Comment by Rab815** [comment]
Still have to use templateUrl in the components

**Comment by Rab815** [comment]
Though I will say the only way to keep it working was with the angular_1_router.js file that came with the repo.  Getting the latest version from npm install @angular/router will not work.

**Comment by petebacondarwin** [comment]
OK, so the problem is that the `dist-tag` for npm is not set correctly...
Install the new router using `npm install @angular/router@0.1.0`

**Comment by Rab815** [comment]
@petebacondarwin  Once you execute the npm install @angular/router@0.1.0, is there a way to verify the version downloaded?  I don't see verson information in the angular_1_router.js file?

**Comment by brandonroberts** [comment]
@Rab815 look in the `node_modules/@angular/router/package.json`

**Comment by Rab815** [comment]
Gotcha!  Thanks!
